File: examples/python/create_vm_test.py
import unittest

from create_vm import (
    NoCompletionResponse, retry_independent, retry_same_operation,
    retry_with_new_operation,
)


class Receiver:
    def __init__(self, protect: bool):
        self.protect = protect
        self.seen: set[str | None] = set()
        self.calls = 0
        self.vms = 0

    def create_vm(self, operation_id: str | None, size: str) -> None:
        self.calls += 1
        if not self.protect or operation_id not in self.seen:
            self.vms += 1
            self.seen.add(operation_id)
        if self.calls == 1:
            raise NoCompletionResponse()


class CreateVmTests(unittest.TestCase):
    def test_weak_repeat_can_create_two(self):
        receiver = Receiver(False)
        retry_independent(receiver, "small")
        self.assertEqual(receiver.vms, 2)

    def test_same_p_protects_one(self):
        receiver = Receiver(True)
        retry_same_operation(receiver, "small", "P")
        self.assertEqual(receiver.vms, 1)

    def test_p_then_q_can_create_two(self):
        receiver = Receiver(True)
        retry_with_new_operation(receiver, "small", "P", "Q")
        self.assertEqual(receiver.vms, 2)


if __name__ == "__main__":
    unittest.main()

File: examples/typescript/create-vm.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { NoCompletionResponse, retryIndependent, retrySameOperation, retryWithNewOperation } from './create-vm.ts';

function receiver(protect: boolean) {
  const seen = new Set<string | null>();
  let calls = 0;
  let vms = 0;
  return {
    async createVM(id: string | null) {
      calls++;
      if (!protect || !seen.has(id)) { vms++; seen.add(id); }
      if (calls === 1) throw new NoCompletionResponse();
    },
    get vms() { return vms; },
  };
}

test('weak repeat can create two VMs', async () => {
  const client = receiver(false);
  await retryIndependent(client, { size: 'small' });
  assert.equal(client.vms, 2);
});
test('same P protects one VM under stronger contract', async () => {
  const client = receiver(true);
  await retrySameOperation(client, { size: 'small' }, 'P');
  assert.equal(client.vms, 1);
});
test('P then Q can create two VMs', async () => {
  const client = receiver(true);
  await retryWithNewOperation(client, { size: 'small' }, 'P', 'Q');
  assert.equal(client.vms, 2);
});
